### ISSUE TITLE
SSTable <name> uses CDCPartitioner partitioner which is different than Murmur3Partitioner partitioner used by the database

### ISSUE THREAD

**Issue by kbr-** [scylladb/scylla#6893]
Steps to reproduce:

1. start node with CDC enabled
2. create a table with CDC enabled
3. insert data into the table
4. nodetool flush
5. restart node

result:
```
ERROR 2020-07-21 14:37:12,462 [shard 0] init - Startup failed: std::runtime_error (Exception while populating keyspace 'ks' with column family 't_scylla_cdc_log' from file 'tmp/ks/t_scylla_cdc_log-d8f601b2cb4e11eaace5000000000000': std::runtime_error (SSTable tmp/ks/t_scylla_cdc_log-d8f601b2cb4e11eaace5000000000000/mc-3-big-Data.db uses com.scylladb.dht.CDCPartitioner partitioner which is different than org.apache.cassandra.dht.Murmur3Partitioner partitioner used by the database))
```
By the way, it's funny that none of our tests catched that (I catched it by accident)

**Comment by haaawk** [comment]
It's interesting because the check seems to be right `v.partitioner.value != to_bytes(_schema->get_partitioner().name()` maybe schema for cdc_log is not persisted correctly.

**Comment by nyh** [comment]
Please add a dtest which reproduces this issue (should be easy - write some data, and reboot the node gently (or flush and stop it abruptly)) so we can test the fix (and be sure the bug never comes back).
Our "longevity" tests should have caught this too - they have various "nemeses", including reboots.  Are we doing any longevity test which includes CDCs?

**Comment by haaawk** [comment]
Interesting observation:

Following sequence fails:

1. start fresh scylla node
2. create cdc table
3. write some rows
4. stop node
5. start node

but if we restart the node after creating the table and before inserting rows everything works:

1. start fresh scylla node
2. create cdc table
3. restart the node
4. write some rows
5. stop node
6. start node

**Comment by haaawk** [comment]
On the same time, partitioner is correctly persisted for cdc table. I also checked that sstables contain the right partitioner in their statistics.

**Comment by haaawk** [comment]
It seems that after all the partitioner is not exactly correctly persisted.

**Comment by haaawk** [comment]
or rather it is persisted correctly but later it's not read for some reason.

**Comment by haaawk** [comment]
It seems that there is a race between features being enabled and users of schema_features.

**Comment by haaawk** [comment]
Schemas for non-system tables are read before gossip is started so no features can be enabled at a time. I guess we could use schema_features::full() for reaching schemas from disk.

**Comment by tgrabiec** [comment]
Is this regression from b97f466438f3e68ba48adafa20d11a26479c7dbb?

Solutions:

  - revert b97f466438f3e68ba48adafa20d11a26479c7dbb and redact schema only when exchanging mutations between nodes, not when loading node's schema

  - Make sure node's features are monotonic across restarts

**Comment by haaawk** [comment]
@tgrabiec that's probably the cause. I don't think reverting it is the solution because probably the commit was introduced for the reason.

I was thinking about making sure that digest calculation uses redaction mechanism but loading schemas for db use to not use it.

**Comment by tgrabiec** [comment]
@haaawk Digest calculation already used redaction before b97f466. b97f466 cares only about applying or exchanging mutations, the node can still load the full schema on boot.

**Comment by haaawk** [comment]
Right I know that digest was using redaction before. What I meant is that the code changed by https://github.com/scylladb/scylla/commit/b97f466438f3e68ba48adafa20d11a26479c7dbb is used in:

```
1. read_table_mutations(..., sstring keyspace_name, sstring table_name, schema_ptr s)
  ^
  +--- db::legacy_schema_migrator::migrator::read_table
2. read_table_mutations(..., const qualified_name& table, schema_ptr s)
  ^
  +--- create_view_from_table_row
  |       ^
  |       +--- create_views_from_schema_partition
  |                 ^
  |                 +--- database::parse_system_tables
  |
  +--- read_tables_for_keyspaces
  |        ^
  |        +--- do_merge_schema
  |
  +--- create_table_from_name
           ^
           +--- create_table_from_table_row
                    ^
                    +--- create_tables_from_tables_partition
                             ^
                             +--- database::parse_system_tables
```

@eliransin please correct me if I'm wrong but I expect that https://github.com/scylladb/scylla/commit/b97f466438f3e68ba48adafa20d11a26479c7dbb was introduced for the do_merge_schema code path. I think usages that originate in database::parse_system_tables shouldn't redact. I'm not sure about db::legacy_schema_migrator::migrator::read_table. What you you think about it @tgrabiec?

So I meant to change that do_merge_schema uses the redaction and database::parse_system_tables doesn't use it.

**Comment by haaawk** [comment]
So I would propose to remove redaction from read_schema_partition_for_table and do it in read_tables_for_keyspaces instead. What do you think?

**Comment by tgrabiec** [comment]
Actually, the intention of b97f466 was to keep the table schema version compatible with the old nodes. So we must redact on boot, otherwise we would have table schema version mismatch and failed writes.

This leaves us with the other option, making sure schema features are monotonic.

**Comment by haaawk** [comment]
I don't understand why the new nodes can't use new schema? As long as they restrain themselves from using new features and "fake" the digest calculation it should be fine, no?

**Comment by haaawk** [comment]
I also don't understand how such new node will switch to usage of the new schema if it has trimmed the schema loaded from disk at startup?

Take this issue as an example.
Then node starts.
It reads schemas from disk before starting any transport.
It trims those schemas because all the features are disabled
It starts transport and gossip
Features get enabled
Schemas stay trimmed for ever

**Comment by haaawk** [comment]
Also if there is a new schema stored on disk that means the appropriate features must have been enabled before. That means making features monotonic and not redacting schemas stored on disk is equivalent, no?

**Comment by haaawk** [comment]
> Also if there is a new schema stored on disk that means the appropriate features must have been enabled before. That means making features monotonic and not redacting schemas stored on disk is equivalent, no?

Actually, on second thought, I don't think this is true because new nodes use new schemas (at least I believe so) but that's ok as long as they don't allow any usage of new columns before appropriate features are enabled.

**Comment by nizarh11** [comment]
The issue is also reproducing on longevity-cdc-100gb-4h-test:

scylla_ami_id= ami-0b7825734a40d9caa
version= 666.development-0.20200723.e7c18963e4b
number_of_nodes= 6

https://jenkins.scylladb.com/job/scylla-master/job/longevity/job/longevity-cdc-100gb-4h-test/64/


```
2020-07-26T12:23:26+00:00  longevity-cdc-100gb-4h-master-db-node-8ee1e712-2 !ERR     | scylla: [shard 0] init - Startup failed: std::runtime_error (Exception while populating keyspace 'cdc_test' with colu
mn family 'test_table_preimage_scylla_cdc_log' from file '/var/lib/scylla/data/cdc_test/test_table_preimage_scylla_cdc_log-5aba2bd2cf3811ea8bdc000000000006': std::runtime_error (SSTable /var/lib/scylla/da
ta/cdc_test/test_table_preimage_scylla_cdc_log-5aba2bd2cf3811ea8bdc000000000006/mc-14-big-Data.db uses com.scylladb.dht.CDCPartitioner partitioner which is different than org.apache.cassandra.dht.Murmur3P
artitioner partitioner used by the database))
```
Logs:
```
+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
|                                                                                              Log links for testrun with test id 8ee1e712-5248-4360-8cc0-ce7e05b0f58f                                                                                               |
+-----------------+-------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Date            | Log type    | Link                                                                                                                                                                                                                               |
+-----------------+-------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| 20190101_010101 | prometheus  | https://cloudius-jenkins-test.s3.amazonaws.com/8ee1e712-5248-4360-8cc0-ce7e05b0f58f/prometheus_snapshot_20200726_123542.tar.gz                                                                                                     |
| 20200726_122651 | grafana     | https://cloudius-jenkins-test.s3.amazonaws.com/8ee1e712-5248-4360-8cc0-ce7e05b0f58f/20200726_122651/grafana-screenshot-overview-20200726_122651-longevity-cdc-100gb-4h-master-monitor-node-8ee1e712-1.png                          |
| 20200726_122651 | grafana     | https://cloudius-jenkins-test.s3.amazonaws.com/8ee1e712-5248-4360-8cc0-ce7e05b0f58f/20200726_122651/grafana-screenshot-scylla-per-server-metrics-nemesis-20200726_123026-longevity-cdc-100gb-4h-master-monitor-node-8ee1e712-1.png |
| 20200726_123555 | grafana     | https://cloudius-jenkins-test.s3.amazonaws.com/8ee1e712-5248-4360-8cc0-ce7e05b0f58f/20200726_123555/grafana-screenshot-overview-20200726_123555-longevity-cdc-100gb-4h-master-monitor-node-8ee1e712-1.png                          |
| 20200726_123555 | grafana     | https://cloudius-jenkins-test.s3.amazonaws.com/8ee1e712-5248-4360-8cc0-ce7e05b0f58f/20200726_123555/grafana-screenshot-scylla-per-server-metrics-nemesis-20200726_123901-longevity-cdc-100gb-4h-master-monitor-node-8ee1e712-1.png |
| 20200726_124429 | db-cluster  | https://cloudius-jenkins-test.s3.amazonaws.com/8ee1e712-5248-4360-8cc0-ce7e05b0f58f/20200726_124429/db-cluster-8ee1e712.zip                                                                                                        |
| 20200726_124429 | loader-set  | https://cloudius-jenkins-test.s3.amazonaws.com/8ee1e712-5248-4360-8cc0-ce7e05b0f58f/20200726_124429/loader-set-8ee1e712.zip                                                                                                        |
| 20200726_124429 | monitor-set | https://cloudius-jenkins-test.s3.amazonaws.com/8ee1e712-5248-4360-8cc0-ce7e05b0f58f/20200726_124429/monitor-set-8ee1e712.zip                                                                                                       |
| 20200726_124429 | sct-runner  | https://cloudius-jenkins-test.s3.amazonaws.com/8ee1e712-5248-4360-8cc0-ce7e05b0f58f/20200726_124429/sct-runner-8ee1e712.zip                                                                                                        |
+-----------------+-------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+

```

**Comment by tgrabiec** [comment]
> I don't understand why the new nodes can't use new schema? As long as they restrain themselves from using new features and "fake" the digest calculation it should be fine, no?

I don't like the idea of solving this problem by faking the digest. It makes the digest wrong, it won't match schema content, which can lead to other bugs where we think the two schemas are the same when in fact they're not. 


> I also don't understand how such new node will switch to usage of the new schema if it has trimmed the schema loaded from disk at startup?

If schema features are monotonic then we don't have to reload the schema when features are enabled. If the feature wasn't enabled it means there couldn't be anything relevant (other than backwards compatible defaults) written to those columns yet. So enabling the feature shouldn't change the value of the schema.

Since schema features are not really monotonic, this doesn't hold, and we'd have to either reload the schema or make sure that schema features are monotonic. I suggested to do the latter. I think it's better because monotonicity makes it easier to reason about and non-monotonic features will cause other feature-specific problems elsewhere.

**Comment by haaawk** [comment]
> > I don't understand why the new nodes can't use new schema? As long as they restrain themselves from using new features and "fake" the digest calculation it should be fine, no?
> 
> I don't like the idea of solving this problem by faking the digest. It makes the digest wrong, it won't match schema content, which can lead to other bugs where we think the two schemas are the same when in fact they're not.

Just to clarify, that was the solution previously, right? New nodes were using the new schema that had to be backward compatible and were only faking the digest. Or was I getting it all wrong?

Are we changing this to make sure that new nodes use old schema until appropriate features are enabled? Have we started any work towards this? I sort of missed the moment this was decided.

> 
> > I also don't understand how such new node will switch to usage of the new schema if it has trimmed the schema loaded from disk at startup?
> 
> If schema features are monotonic then we don't have to reload the schema when features are enabled. If the feature wasn't enabled it means there couldn't be anything relevant (other than backwards compatible defaults) written to those columns yet. So enabling the feature shouldn't change the value of the schema.
> 
> Since schema features are not really monotonic, this doesn't hold, and we'd have to either reload the schema or make sure that schema features are monotonic. I suggested to do the latter. I think it's better because monotonicity makes it easier to reason about and non-monotonic features will cause other feature-specific problems elsewhere.

I guess we could try to make schema features monotonic but what about this schema reload? Won't that create races? Once a feature is enabled cluster starts to serve requests with that new feature. At the same time, the schema will be reloaded but not instantly. How would we ensure this is not happening? Reloading schema before enabling feature? Then schema would be redacted during loading again.

**Comment by tgrabiec** [comment]
> > > I don't understand why the new nodes can't use new schema? As long as they restrain themselves from using new features and "fake" the digest calculation it should be fine, no?
> > 
> > 
> > I don't like the idea of solving this problem by faking the digest. It makes the digest wrong, it won't match schema content, which can lead to other bugs where we think the two schemas are the same when in fact they're not.
> 
> Just to clarify, that was the solution previously, right? New nodes were using the new schema that had to be backward compatible and were only faking the digest. Or was I getting it all wrong?

Maybe part of the confusion is that the term "digest" is ambiguous. There is a per-node schema digest, which is a digest of the whole schema, and there's "table schema version", which is a digest of the schema of a single table. 

The fix in b97f466 deals with table schema version (also called digest in the commit message), by redacting mutations before it is calculated. Table schema version was not faked before the fix, the per-node schema version was.

> 
> Are we changing this to make sure that new nodes use old schema until appropriate features are enabled? Have we started any work towards this? I sort of missed the moment this was decided.
> 
> > > I also don't understand how such new node will switch to usage of the new schema if it has trimmed the schema loaded from disk at startup?
> > 
> > 
> > If schema features are monotonic then we don't have to reload the schema when features are enabled. If the feature wasn't enabled it means there couldn't be anything relevant (other than backwards compatible defaults) written to those columns yet. So enabling the feature shouldn't change the value of the schema.
> > Since schema features are not really monotonic, this doesn't hold, and we'd have to either reload the schema or make sure that schema features are monotonic. I suggested to do the latter. I think it's better because monotonicity makes it easier to reason about and non-monotonic features will cause other feature-specific problems elsewhere.
> 
> I guess we could try to make schema features monotonic but what about this schema reload? Won't that create races? Once a feature is enabled cluster starts to serve requests with that new feature. At the same time, the schema will be reloaded but not instantly. How would we ensure this is not happening? Reloading schema before enabling feature? Then schema would be redacted during loading again.

There is a problem here indeed. Suppose node A has the feature enabled but node B doesn't have it enabled yet. Node A will send requests referencing the new (unredacted) table schema version. Node B will receive the mutations and will be able to apply them, but will still redact them when materializing the schema object. As a result the table schema version requesting will fail, since we'll materialize the wrong schema version.

I now think that b97f466 is not the right way to fix the problem it aimed at fixing. Instead, schema change commands should not be setting any columns which are not supported yet, so that this redacting is not needed. Lack of the value in the new column should be interpreted as if the default value was there. @eliransin, will that fix your bug as well?

**Comment by haaawk** [comment]
@tgrabiec @eliransin Maybe this would solve the problem -> https://github.com/scylladb/scylla/pull/6943

**Comment by tgrabiec** [comment]
@haaawk Redaction on merge needs to be consistent with redaction on boot, since even during rolling upgrade, nodes can reboot,  and we should end up with the same decision about digests.

**Comment by nyh** [comment]
This fix reverts commit b97f466438f3e68ba48adafa20d11a26479c7dbb that isn't in 4.2, so no need to backport the fix.
